Add JVM options configuration file

jasontedor opened the pull request:
This commit adds a new configuration file jvm.options to centralize and
simplify management of JVM options. This separates the configuration of
the JVM from the packaging scripts (bin/elasticsearch*, bin/service.bat,
and init.d/elasticsearch) simplifying end-user operational management of
custom JVM options.

Closes #17121

jasontedor committed:
Add JVM options configuration file

This commit adds a new configuration file jvm.options to centralize and
simplify management of JVM options. This separates the configuration of
the JVM from the packaging scripts (bin/elasticsearch*, bin/service.bat,
and init.d/elasticsearch) simplifying end-user operational management of
custom JVM options.

dakrone commented on distribution/src/main/resources/bin/elasticsearch:
```diff
@@ -44,6 +44,13 @@
 
 # Check to see if you are trying to run this without building it first. Gradle
```
This comment is now in the wrong place, should be after the `jvm_options` function

jasontedor replied:
@dakrone I pushed 00033b3da39ecd841d3e7ef8553b90254f8b93a1.

dakrone commented on distribution/src/main/resources/bin/elasticsearch:
```diff
@@ -74,6 +81,17 @@ ES_HOME=`dirname "$SCRIPT"`/..
 ES_HOME=`cd "$ES_HOME"; pwd`
 
+if [ -z "$ES_JVM_OPTIONS" ]; then
+    for jvm_options in "$ES_HOME"/config/jvm.options \
```
This shadows the function defined above :(

jasontedor replied:
@dakrone I pushed 9460289bee2cda9cb91b2f9d6370663f53a00abb.

dakrone commented on distribution/src/main/resources/bin/elasticsearch:
```diff
@@ -126,10 +144,10 @@ export HOSTNAME
 daemonized=`echo $* | egrep -- '(^-d |-d$| -d |--daemonize$|--daemonize )'`
 if [ -z "$daemonized" ] ; then
-    exec "$JAVA" $JAVA_OPTS $ES_JAVA_OPTS -Des.path.home="$ES_HOME" -cp "$ES_CLASSPATH" \
+    exec "$JAVA" $ES_JAVA_OPTS -Des.path.home="$ES_HOME" -cp "$ES_CLASSPATH" \
```
Shouldn't this be `-Ees.path.home` for the new setup?

jasontedor replied:
No, this is being passed as an argument to the JVM to set this as a system property (note that it comes before the class name, it's not an argument to Elasticsearch).

dakrone replied:
Ahh okay

dakrone commented on distribution/src/main/resources/bin/elasticsearch.bat:
```diff
@@ -35,6 +35,14 @@ FOR /F "usebackq tokens=1* delims= " %%A IN (!params!) DO (
 SET HOSTNAME=%COMPUTERNAME%
 
+if "%ES_JVM_OPTIONS%" == "" (
+set ES_JVM_OPTIONS="%~dp0\..\config\jvm.options"
```
I think a comment explaining `%~dp0` would be great here :)

jasontedor replied:
@dakrone I pushed 2b1c191b121c1e5ecdec329639625bb701beb7b1.

dakrone commented on distribution/src/main/resources/bin/elasticsearch.bat:
```diff
@@ -35,6 +35,14 @@ FOR /F "usebackq tokens=1* delims= " %%A IN (!params!) DO (
+)
+
+@setlocal
+for /F "usebackq delims=" %%a in (`findstr /b \- "%ES_JVM_OPTIONS%"`) do set JVM_OPTIONS=!JVM_OPTIONS! %%a
```
Same here, explaining what this is doing

jasontedor replied:
@dakrone I pushed 2b1c191b121c1e5ecdec329639625bb701beb7b1.

dakrone commented:
I'm sad that this removes support for the `JAVA_OPTS` env variable, since that is a standard for most JVM software and it'd be nice to keep in line with all the other java servers that do support it.

dakrone commented:
Also, you're going to hate me for this, but I find the changes to the Windows stuff to be suuuper dangerous, I figure there isn't a great way to vagrant test those, but is there some way we can test this stuff in CI?

jasontedor commented:
> I'm sad that this removes support for the `JAVA_OPTS` env variable, since that is a standard for most JVM software and it'd be nice to keep in line with all the other java servers that do support it.

It's not standard. Please see my comments on #17121 justifying removing `JAVA_OPTS`.

jasontedor commented:
> Also, you're going to hate me for this, but I find the changes to the Windows stuff to be suuuper dangerous, I figure there isn't a great way to vagrant test those, but is there some way we can test this stuff in CI?

The `elasticsearch.bat` script will be implicitly tested on Windows when we start up external nodes for integration tests. I also tested it manually.

I manually tested the service.bat extensively, which I think that is all that we can do right now. My understanding is that there are plans to put a different service installer in place so I don't think we should invest any effort into making the current one testable under CI.

dakrone commented:
> It's not standard. Please see my comments on #17121 justifying removing JAVA_OPTS.

Okay, since you're removing it, what do you think of the warning Clint mentioned if it was set and ES_JAVA_OPTS is not set?

jasontedor committed:
Fix placement of comment in bin/elasticsearch

This commit fixes the placement of a comment in bin/elasticsearch. The
comment was made out of place by the addition of a function definition
but order is restored.

jasontedor committed:
Rename function that parses JVM options

This commit renames the function that parses JVM options in
bin/elasticsearch from jvm_options to parse_jvm_options. The reason for
the rename is because a for-loop variable was shadowing the name of this
function and changing the function name further clarifies the purpose of
the function.

jasontedor committed:
Warn in bin/elasticsearch if JAVA_OPTS is set

JAVA_OPTS is not a built-in mechanism for passing options to the JVM but
many people think that it is. We do not respect JAVA_OPTS, but this
commit adds a warning if it is set in case the end-user thinks that it
will affect Elasticsearch.

dakrone commented on distribution/src/main/resources/bin/elasticsearch:
```diff
@@ -53,6 +54,12 @@ EOF
 
+parse_jvm_options() {
+  if [ -f "$1" ]; then
+    echo "$(grep "^-" "$1" | tr '\n' ' ')"
```
Suuuper minor nit, could be `grep -F`

jasontedor replied:
@dakrone That would cause `grep` to interpret the pattern as a fixed string, not as a regular expression but I'm using `^` as the beginning-of-line metacharacter and not as the literal `^`.

dakrone replied:
Oh duh, of course it wouldn't work, I wasn't thinking properly

jasontedor commented:
> Okay, since you're removing it, what do you think of the warning Clint mentioned if it was set and ES_JAVA_OPTS is not set?

@dakrone I pushed 7743c422dbd14bcebdfbe1d253bf2d4e20445689 to warn about `JAVA_OPTS`

dakrone commented on distribution/src/main/resources/bin/elasticsearch:
```diff
@@ -74,6 +81,17 @@ ES_HOME=`dirname "$SCRIPT"`/..
+    done
+fi
+
+ES_JAVA_OPTS="$(parse_jvm_options "$ES_JVM_OPTIONS") $ES_JAVA_OPTS"
```
I think the precedence should be documented at the top of the file also?

jasontedor replied:
@dakrone I've gone back and forth on this a few times in my mind. Some JVM conflicting options will cause the JVM to not start, and some are last-write wins. I don't think that people should be relying on this *ever*, and I so I don't think that it should be documented. Do you still think it should be documented?

dakrone commented on distribution/src/main/resources/bin/elasticsearch:
```diff
@@ -114,7 +132,7 @@ fi
 # works around https://bugs.launchpad.net/ubuntu/+source/jayatana/+bug/1441487
 if [ "x$JAVA_TOOL_OPTIONS" != "x" ]; then
```
Could probably just duplicate this for `JAVA_OPTS`

jasontedor replied:
I think we raced, but I pushed 7743c422dbd14bcebdfbe1d253bf2d4e20445689.

jasontedor committed:
Add comments to JVM options parsing for Windows

This commit adds some comments to the portion of bin/elasticsearch.bat
that is responsible for parsing the JVM options.

jasontedor commented:
@dakrone Thanks for the review; I responded to all of your feedback. 😄

kimchy commented:
@jasontedor when someone upgrades to 5.0 with this change, and they rely on the environment variables, they will get a silent startup with wrong settings, right? I am concerned about that, I imagine someone upgrading to 5.0, not understanding why the system fails, and taking time to figure out that a system property no longer applies. Maybe we can log a warning or not startup if this is set, at least for 5.x?

kimchy commented:
also, for a separate issue, I always wished we could, maybe in certain OSes, figure out the memory allocated, and set the defaults OOB to be the one we typically recommend (less than 26gb, 50% of OS).

jasontedor commented:
> Maybe we can log a warning or not startup if this is set, at least for 5.x?

@kimchy Assuming this PR is integrated, my plan was to open a second PR for 2.x to add warnings for all the environment variables that are being removed here but I think that is a very good idea. I'll push another commit soon.

jasontedor commented:
> also, for a separate issue, I always wished we could, maybe in certain OSes, figure out the memory allocated, and set the defaults OOB to be the one we typically recommend (less than 26gb, 50% of OS).

Yes, this is something that we can do. I opened #17686.

jasontedor committed:
Fail if unsupported environment variables are set

This commit adds a hard failure on startup if any unsupported
environment variables from older versions of Elasticsearch are still
set.

jasontedor commented:
@kimchy I pushed d4cf0f021a28544700af543299d878cec8db4438 to fail if any of the now-unsupported environment variables are set.

kimchy commented:
@jasontedor awesome, thanks!

kimchy commented:
@jasontedor I had another thought, I am on the fence on it, so would love your opinion. This is a change where someone has to make for 5.0, without an opportunity to have both options (the new one, using the right way to set it, and the old way, being deprecated). What do you think about "converting" the old env variables to `ES_JAVA_OPTS`, and add a deprecation warning for 5.x? Is it even possible (I am thinking of things like collisions between conflicting settings, in which case we just need to make this change as it is)?

jasontedor commented:
> I am thinking of things like collisions between conflicting settings, in which case we just need to make this change as it is

@kimchy It's exactly because of this that I think that we need to take the change as is. Either:
 - the user will have conflicting settings which cause the JVM to fail to start in which case they will be forced to convert to the new way
 - the user will have conflicting settings which do not cause the JVM to fail to start, but we will be silently ignoring one of their settings (e.g., imagine a user that has set `ES_HEAP_SIZE`, we convert it to `ES_JAVA_OPTS` and then they later modify the heap size setting in `jvm.options`: then currently their old `ES_HEAP_SIZE` settings will still take precedence); for these users the best option is also to convert to the new way

I do not think there is a *nice* way for the two ways to sit side-by-side.

kimchy commented:
> I do not think there is a nice way for the two ways to sit side-by-side.

@jasontedor aye, I hear you... . I suspect many users will fail when they upgrade to 5.0, and need to change to use the new mechanism. Not a big change, but still there. Can we maybe give them examples in the error message, specifically around the most common setting which I suspect is setting the heap size (i.e. show an example how to set it using ES_JAVA_OPTS).

jasontedor committed:
Add migration messages when rejecting env. vars

This commit adds migration messages to the startup scripts when
rejecting the old unsupported environment variables.

jasontedor commented:
@kimchy I pushed 606212956aa3c30dfeaedd7b696cdb883ad5ead4. This provides error messages like this on startup:

```
$ ES_HEAP_SIZE=4g ES_DIRECT_SIZE=1g ~/elasticsearch/elasticsearch-5.0.0-alpha1-SNAPSHOT/bin/elasticsearch
Error: encountered environment variables that are no longer supported
Use jvm.options or ES_JAVA_OPTS to configure the JVM
ES_HEAP_SIZE=4g: set -Xms4g and -Xmx4g in jvm.options or add "-Xms4g -Xmx4g" to ES_JAVA_OPTS
ES_DIRECT_SIZE=1g: set -XX:MaxDirectMemorySize=1g in jvm.options or add "-XX:MaxDirectMemorySize=1g" to ES_JAVA_OPTS
```

kimchy commented:
@jasontedor you rock, thanks!

dakrone commented:
LGTM

jasontedor committed:
Additional handling of ES_GC_LOG_FILE

This commit adds handling of the old ES_GC_LOG_FILE environment variable
to treat it like the other removed environment variables.

jasontedor committed:
Rewrite if-statement for shell compatibility

This commit rewrites an if-statement in bin/elasticsearch for
compatibility with some shells (e.g., dash on Ubuntu).

jasontedor committed:
Remove unsupported env. var from node spinup

This commit removes setting ES_GC_OPTS from NodeInfo.groovy as this
environment variable is no longer supported.

jasontedor commented:
Thanks for a patient and helpful review @dakrone and great feedback @kimchy!
